Source/WebCore/dom/SlotAssignment.cpp

```diff
@@ -441,15 +452,21 @@ void NamedSlotAssignment::assignSlots(ShadowRoot& shadowRoot)
 void NamedSlotAssignment::assignToSlot(Node& child, const AtomString& slotName)
 {
     ASSERT(!slotName.isNull());
+    std::optional<Style::PseudoClassChangeInvalidation> styleInvalidation;
     if (slotName == defaultSlotName()) {
         auto defaultSlotEntry = m_slots.find(defaultSlotName());
-        if (defaultSlotEntry != m_slots.end())
+        if (defaultSlotEntry != m_slots.end()) {
             defaultSlotEntry->value->assignedNodes.append(child);
+            if (auto* slotElement = defaultSlotEntry->value->element.get())
+                styleInvalidation.emplace(*slotElement, CSSSelector::PseudoClass::HasSlotted, slotElement->hasFlattenedSlottedContent());
```

Comment:
This is too late. `NamedSlotAssignment::assignSlots` is only called lazily.
We need to detect when `:has-slotted` is used and eagerly resolve the assignment.
See how `shadowRoot.shouldFireSlotchangeEvent() && hasAssignedNodes(shadowRoot, slot)` is checked to enqueue `slotchange` event.